sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -133,12 +133,48 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
 
   private def visitCompoundBodyImpl(
       ctx: CompoundBodyContext,
-      label: Option[String]): CompoundBody = {
+      label: Option[String],
+      allowVarDeclare: Boolean): CompoundBody = {
     val buff = ListBuffer[CompoundPlanStatement]()
     ctx.compoundStatements.forEach(compoundStatement => {
       buff += visit(compoundStatement).asInstanceOf[CompoundPlanStatement]
     })
 
+    val compoundStatements = buff.toList
+
+    if (allowVarDeclare) {
+      val declareVarStatement = compoundStatements
+        .dropWhile(statement => statement.isInstanceOf[SingleStatement] &&
+          statement.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
+        .filter(_.isInstanceOf[SingleStatement])
+        .find(_.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
+
+      declareVarStatement match {
+        case Some(SingleStatement(parsedPlan)) =>
+          throw SqlScriptingErrors.variableDeclarationOnlyAtBeginning(
+            parsedPlan.asInstanceOf[CreateVariable]
+              .name.asInstanceOf[UnresolvedIdentifier]
+              .nameParts.last,
+            parsedPlan.origin.line.get.toString)
+        case _ =>
+      }
+
+    } else {
+      val declareVarStatement = compoundStatements
+        .filter(_.isInstanceOf[SingleStatement])
+        .find(_.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
```

Comment:
A better way to avoid duplicated code
```
val candidates = if (allowVarDeclare) {
  compoundStatements.dropWhile
} else {
  compoundStatements
}
val varNames = candidates.filter...
...
```

Reply:
Done